virtio-devices/src/mem.rs

```diff
@@ -538,6 +558,7 @@ impl MemEpollHandler {
                                     host_addr: self.host_addr,
                                     host_fd: self.host_fd,
                                     plug: true,
+                                    hugepages: &mut self.hugepages,
```

Comment:
Don't do this. Make `virtio_mem_state_change_request` and the other functions methods (i.e. pass in `&self`)

Reply:
```
error[E0499]: cannot borrow `*self` as mutable more than once at a time
   --> virtio-devices/src/mem.rs:535:29
    |
524 |         for avail_desc in self.queue.iter(&mem) {
    |                           ---------------------
    |                           |
    |                           first mutable borrow occurs here
    |                           first borrow later used here

```
Got this error.  Do you have any idea on this issue?

Best,
Hui

Reply:
I think this is because `self.queue.iter()` takes a mutable reference to the queue, which further turns into a mutable borrowed `self`.

Arguably you're trying to change different parts of `self`, but Rust doesn't seem to understand that.

You may be able to use `std::mem::replace` to yank the queue out from `self` before processing then put it back after the loop.

See https://github.com/liuw/cloud-hypervisor/commit/0f98721ac7ab3564e3d96cd8b790dd85f02c0b21. It does get pass your original error, but now it is complaining about something due to more or less the same reason.

Obligatory warning: I have no idea if my code is correct, so don't use it as-is. I'm just curious about the reasoning behind these errors rather than virtio-mem itself.

Reply:
Another possibility is to take advantage of interior mutability. Maybe `RefCell` can be useful. With that you can take immutable references to `self` in all associated functions and then use `borrow_mut()` to get a mutable reference to `mem_state` and `hugepages`.

Then again, this is just my random thought. ;-)

Reply:
@liuw Thanks!
I think it will be great if you can send a special PR to fix it.